llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -2888,25 +2885,19 @@ static const SCEV *getStrideFromPointer(Value *Ptr, ScalarEvolution *SE, Loop *L
     return nullptr;
 
   V = S->getStepRecurrence(*SE);
-  if (!V)
-    return nullptr;
 
   // Strip off the size of access multiplication if we are still analyzing the
   // pointer.
   if (OrigPtr == Ptr) {
-    if (const SCEVMulExpr *M = dyn_cast<SCEVMulExpr>(V)) {
-      if (M->getOperand(0)->getSCEVType() != scConstant)
+    if (auto *M = dyn_cast<SCEVMulExpr>(V)) {
+      auto *StepConst = dyn_cast<SCEVConstant>(M->getOperand(0));
+      if (!StepConst)
         return nullptr;
 
-      const APInt &APStepVal = cast<SCEVConstant>(M->getOperand(0))->getAPInt();
-
-      // Huge step value - give up.
-      if (APStepVal.getBitWidth() > 64)
+      if (auto StepVal = StepConst->getAPInt().trySExtValue();
+          !StepVal || StepVal != 1)
         return nullptr;
 
-      int64_t StepVal = APStepVal.getSExtValue();
-      if (PtrAccessSize != StepVal)
```

Comment:
What happened to this check? Don't you still need this

`if (PtrAccessSize != *StepVal)`

If the tests are passing with this check removed it sounds like we're missing a test somewhere.

Reply:
See `StepVal != 1`

Reply:
OK I see that `PtrAccessSize` seems to be always set to 1. It wasn't immediately obvious to me that this was a fixed value. In that case, shouldn't this be `*StepVal != 1`? It's an optional value so you have to dereference it.

Reply:
It looks to me like `StepVal != 1` is logically equivalent to `!StepVal` because StepVal is an optional and will get cast to a boolean so essentially this comparison becomes `true != 1`.

Reply:
Ah ok, fair enough! I learned something new today. :)